hibernate-core/src/main/java/org/hibernate/loader/ast/internal/AbstractMultiNaturalIdLoader.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.loader.ast.internal;
+
+
+import org.hibernate.LockOptions;
+import org.hibernate.engine.spi.EntityEntry;
+import org.hibernate.engine.spi.EntityKey;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.loader.ast.spi.MultiNaturalIdLoadOptions;
+import org.hibernate.loader.ast.spi.MultiNaturalIdLoader;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Jan Schatteman
+ */
+public abstract class AbstractMultiNaturalIdLoader<E> implements MultiNaturalIdLoader<E> {
+	private final EntityMappingType entityDescriptor;
+
+	protected MultiNaturalIdLoadOptions options;
+
+	public AbstractMultiNaturalIdLoader(EntityMappingType entityDescriptor) {
+		this.entityDescriptor = entityDescriptor;
+	}
+
+	@Override
+	public <K> List<E> multiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		assert naturalIds != null;
+
+		this.options = options;
+
+		if ( naturalIds.length == 0 ) {
+			return Collections.emptyList();
+		}
+
+		return options.isOrderReturnEnabled()
+				? performOrderedMultiLoad( naturalIds, options, session )
+				: performUnorderedMultiLoad( naturalIds, options, session );
+	}
+
+	private <K> List<E> performUnorderedMultiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		if ( MultiKeyLoadLogging.MULTI_KEY_LOAD_LOGGER.isTraceEnabled() ) {
+			MultiKeyLoadLogging.MULTI_KEY_LOAD_LOGGER.tracef( "Unordered MultiLoad Starting - `%s`", getEntityDescriptor().getEntityName() );
+		}
+
+		return unOrderedMultiLoad(
+				naturalIds,
+				session,
+				options.getLockOptions() == null ? LockOptions.NONE : options.getLockOptions()
+		);
+	}
+
+	protected abstract <K> List<E> unOrderedMultiLoad(K[] naturalIds, SharedSessionContractImplementor session, LockOptions lockOptions);
+
+	private <K> List<E> performOrderedMultiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		if ( MultiKeyLoadLogging.MULTI_KEY_LOAD_LOGGER.isTraceEnabled() ) {
+			MultiKeyLoadLogging.MULTI_KEY_LOAD_LOGGER.tracef( "Ordered MultiLoad Starting - `%s`", getEntityDescriptor().getEntityName() );
+		}
+
+		return orderedMultiLoad(
+				naturalIds,
+				session,
+				options.getLockOptions() == null ? LockOptions.NONE : options.getLockOptions()
+		);
+	}
+
+	protected <K> List<E> orderedMultiLoad( K[] naturalIds, SharedSessionContractImplementor session, LockOptions lockOptions ) {
+
+		unOrderedMultiLoad( naturalIds, session, lockOptions );
```

Comment:
"Unordered" is a word you can easily find in the dictionary, so `unorderedMultiLoad`. :-)

Reply:
yep, sorry about that, slipped by me :blush: